version/src/main/java/io/smallrye/common/version/VersionRange.java

```diff
@@ -0,0 +1,69 @@
+package io.smallrye.common.version;
+
+import static io.smallrye.common.version.Messages.msg;
+
+import java.util.function.Predicate;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * A {@link VersionRange} is a predicate that tests if a version string is within a specified range.
+ */
+public class VersionRange implements Predicate<String> {
+
+    /**
+     * Range in format "[1.0,2.0)"
+     */
+    private static final Pattern RANGE_PATTERN = Pattern.compile("([\\[\\(])(.*),(.*)([\\]\\)])");
+
+    private final VersionScheme versionScheme;
+    private final Bound lowerBound;
+    private final Bound upperBound;
+
+    public VersionRange(VersionScheme versionScheme, String rangePattern) {
```

Comment:
This might be nicer if it were a default method on `VersionScheme`. And, you could even make it use composable sub-predicates for arbitrary ranges, like this:

```java
public interface VersionScheme {
    // ...
    default boolean lt(String base, String other) {
        return compare(base, other) < 0;
    }
    // and le, gt, ge

    default Predicate<String> whenLt(String base) {
        return other -> lt(base, other);
    }
    // and whenLe, whenGt, whenGe
```

Then you can easily compose:

```java
    VersionScheme vs = VersionScheme.MAVEN;
    Predicate<String> isVersion1 = vs.whenGe("1").and(vs.whenLt("2"));
```

And then it's pretty trivial to convert the regex pattern result directly into predicates using this API.

Reply:
I'd leave that as an exercise for the reader 😉

Reply:
I prefer it to be solved this way, please.

Reply:
Introduced in [`950941d` (#330)](https://github.com/smallrye/smallrye-common/pull/330/commits/950941db041b72b1a0e97f95f972b7d36a52b62e)